runtime/lua/vim/treesitter/_fold.lua

```diff
@@ -361,4 +361,93 @@ function M.foldexpr(lnum)
   return foldinfos[bufnr].levels[lnum] or '0'
 end
 
+---@package
+---@return { [1]: string, [2]: string[] }[]|string|nil
+function M.foldtext()
+  local foldstart = vim.v.foldstart
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@type LanguageTree|nil
+  local parser = vim.F.npcall(ts.get_parser, bufnr)
+  if not parser then
+    return vim.fn.foldtext()
+  end
+
+  local query = ts.query.get(parser:lang(), 'highlights')
+  if not query then
+    return vim.fn.foldtext()
+  end
+
+  local tree = parser:parse({ foldstart - 1, foldstart })[1]
+
+  local line = api.nvim_buf_get_lines(bufnr, foldstart - 1, foldstart, false)[1]
+
+  ---@type { [1]: string, [2]: string[], range: { [1]: integer, [2]: integer } }[] | { [1]: string, [2]: string[] }[]
+  local result = {}
+
+  local line_pos = 0
+
+  for id, node, metadata in query:iter_captures(tree:root(), 0, foldstart - 1, foldstart) do
+    local name = query.captures[id]
+    local start_row, start_col, end_row, end_col = node:range()
+
+    local priority = tonumber(metadata.priority or vim.highlight.priorities.treesitter)
+
+    if start_row == foldstart - 1 and end_row == foldstart - 1 then
+      -- check for characters ignored by treesitter
+      if start_col > line_pos then
+        table.insert(result, {
+          line:sub(line_pos + 1, start_col),
+          { { 'Folded', priority } },
+          range = { line_pos, start_col },
+        })
+      end
+      line_pos = end_col
+
+      local text = ts.get_node_text(node, 0)
```

Comment:
Is it possible to use `line` here instead of calling `get_node_text`? We already called `nvim_buf_get_lines` to get buffer text, this will do the same thing so is redundant and possibly wasteful.

If this correctly handles an edge case then it's ok to leave as-is, but please leave a comment indicating that.

Reply:
No edgecase that I can think of, I'll change it to use `line:sub`.